subprojects/execution/src/main/java/org/gradle/internal/execution/caching/impl/DefaultCachingStateBuilder.java

```diff
@@ -0,0 +1,326 @@
+    private final ImmutableSortedMap.Builder<String, String> nonCacheableInputPropertiesBuilder = ImmutableSortedMap.naturalOrder();
+    private ImmutableSortedSet<String> outputProperties = ImmutableSortedSet.of();
+    private final ImmutableList.Builder<CachingDisabledReason> noCachingReasonsBuilder = ImmutableList.builder();
+
+    @Override
+    public final void withImplementation(ImplementationSnapshot implementation) {
+        this.implementation = implementation;
+        processImplementation(implementation);
+    }
+
+    protected void processImplementation(ImplementationSnapshot implementation) {
+        if (implementation.isUnknown()) {
+            noCachingReasonsBuilder.add(new CachingDisabledReason(
+                NON_CACHEABLE_IMPLEMENTATION,
+                "Implementation type " + implementation.getUnknownReason()
+            ));
+        }
+    }
+
+    @Override
+    public final void withAdditionalImplementations(Iterable<ImplementationSnapshot> additionalImplementations) {
+        this.additionalImplementations = ImmutableList.copyOf(additionalImplementations);
+        for (ImplementationSnapshot additionalImplementation : additionalImplementations) {
+            processAdditionalImplementation(additionalImplementation);
+        }
+    }
+
+    protected void processAdditionalImplementation(ImplementationSnapshot additionalImplementation) {
+        if (additionalImplementation.isUnknown()) {
+            noCachingReasonsBuilder.add(new CachingDisabledReason(
+                NON_CACHEABLE_ADDITIONAL_IMPLEMENTATION,
+                "Additional implementation type " + additionalImplementation.getUnknownReason()
+            ));
+        }
+    }
+
+    @Override
+    public final void withInputValueFingerprints(Map<String, ValueSnapshot> fingerprints) {
+        fingerprints.forEach((propertyName, fingerprint) -> {
+            Hasher hasher = Hashing.newHasher();
+            fingerprint.appendToHasher(hasher);
+            if (hasher.isValid()) {
+                HashCode hash = hasher.hash();
+                recordInputValueFingerprint(propertyName, hash);
+            } else {
+                markInputValuePropertyNotCacheable(propertyName, hasher.getInvalidReason());
+            }
+        });
+    }
+
+    protected void recordInputValueFingerprint(String propertyName, HashCode fingerprint) {
+        inputValueFingerprintsBuilder.put(propertyName, fingerprint);
+    }
+
+    protected void markInputValuePropertyNotCacheable(String propertyName, String nonCacheableReason) {
+        nonCacheableInputPropertiesBuilder.put(propertyName, nonCacheableReason);
+    }
+
+    @Override
+    public final void withInputFilePropertyFingerprints(Map<String, CurrentFileCollectionFingerprint> fingerprints) {
+        this.inputFileFingerprints = ImmutableSortedMap.copyOf(fingerprints);
+        fingerprints.forEach((propertyName, fingerprint) -> processInputFileFingerprint(propertyName, fingerprint));
```

Comment:
Can we override the current method and add this code to the subclass? So we don't need to iterate when we are not logging the hashes?

Reply:
Done.